### PR TITLE
release/public-v1: port to gaea

### PR DESCRIPTION
**Pull request by climbfuji**
## Desription

Changes required to port to gaea:
- Added support for using clusters (an alternative way to partitions, and used on gaea) with SLURM in the regional workflow and rocoto
- Update several configuration scripts with changes for Gaea
- Add `ush/config.gaea.sh` for reference - can be deleted if considered unnecessary

Additonal changes:
- fixes a bug in `scripts/exregional_make_grid.sh`

Associated PRs:

https://github.com/NOAA-EMC/regional_workflow/pull/373
https://github.com/ufs-community/ufs-srweather-app/pull/69

## TESTS CONDUCTED: 

See https://github.com/ufs-community/ufs-srweather-app/pull/69

## ISSUE (optional): 

Fixes https://github.com/NOAA-EMC/regional_workflow/issues/372

## CONTRIBUTORS (optional): 

@JeffBeck-NOAA

### CODE REVIEW

**Comment by JeffBeck-NOAA** [comment]
@fentoad72, can you please test these changes in the workflow on Gaea through use of [PR #69](https://github.com/ufs-community/ufs-srweather-app/pull/69) in the ufs-srweather-app?  Thanks!

File: ush/config_defaults.sh [hunk 1]
```diff
@@ -61,6 +61,17 @@ RUN_ENVIR="nco"
 # machine-dependent value.  This is not used if SCHED is not set to 
 # "slurm".
 #
+# CLUSTERS_DEFAULT:
+# BLA BLA
```
Comment (gsketefian) [inline ush/config_defaults.sh]: Needs proper documentation.

File: ush/config_defaults.sh [hunk 2]
```diff
@@ -78,6 +89,15 @@ RUN_ENVIR="nco"
 # or is set to an empty string, it will be (re)set to a machine-dependent 
 # value.  This is not used if SCHED is not set to "slurm".
 #
+# CLUSTERS_HPSS:
+# BLA BLA
```
Comment (gsketefian) [inline ush/config_defaults.sh]: Fix documentation.

File: ush/config_defaults.sh [hunk 3]
```diff
@@ -92,6 +112,14 @@ RUN_ENVIR="nco"
 # machine-dependent value.  This is not used if SCHED is not set to 
 # "slurm".
 #
+# CLUSTERS_FCST:
+# BLABLA
```
Comment (gsketefian) [inline ush/config_defaults.sh]: Fix documentation.

File: ush/setup.sh [hunk 4]
```diff
@@ -439,6 +439,20 @@ case $MACHINE in
     QUEUE_FCST=${QUEUE_FCST:-"normal"}
     ;;
 
+  "GAEA")
+    NCORES_PER_NODE=36
+    SCHED="slurm"
+    #PARTITION_DEFAULT=${PARTITION_DEFAULT:-"c4"}
```
Comment (gsketefian) [inline ush/setup.sh]: Can remove commented out PARTITION lines.

File: ush/config.gaea.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
+MACHINE="gaea"
```
Comment (gsketefian) [inline ush/config.gaea.sh]: In the regional_workflow/ush directory, we'd like to only have the two platform-independent tests config.community.sh and config.nco.sh.  The full set of tests is in regional_workflow/tests/baseline_configs.  There is already a test for the graduate student test in that directory, in the file config.GST_release_public_v1.sh.  That can be run using the regional_workflow/tests/run_experiments.sh script (that script will build the full configuration file for this case; we'll have to add the GAEA case in several places in that script).  I can walk you through how to run that script if you like.

File: ush/setup.sh (hunk 4 above)
Comment (climbfuji) [reply ush/setup.sh]: Done.

File: ush/config_defaults.sh (hunk 2 above)
Comment (climbfuji) [reply ush/config_defaults.sh]: Done.

File: ush/config_defaults.sh (hunk 3 above)
Comment (climbfuji) [reply ush/config_defaults.sh]: Done.

File: ush/config_defaults.sh (hunk 1 above)
Comment (climbfuji) [reply ush/config_defaults.sh]: Done.

File: ush/config.gaea.sh (hunk 5 above)
Comment (climbfuji) [reply ush/config.gaea.sh]: I see two places that define certain paths for JET, I assume we just need to do the same for GAEA. I probably need to coordinate this with @fentoad72 who is managing staging the data for the SRW App.

Comment (climbfuji) [reply ush/config.gaea.sh]: Tried:
```
./run_experiments.sh expts_file="baseline_configs/config.GST_release_public_v1.sh" machine="GAEA" account="esrl_bmcs" 2>&1 | tee run_experiments.log
```
Here is the error I got:
```
After processing, the number of experiments to run (num_expts) is:
  num_expts = 17
The list of forecast experiments to run (one experiment per line) is gi-
ven by:
  'RUN_ENVIR="community"'
  'PREEXISTING_DIR_METHOD="rename"'
  'PREDEF_GRID_NAME="RRFS_CONUS_25km"'
  'GRID_GEN_METHOD="ESGgrid"'
  'QUILTING="TRUE"'
  'CCPP_PHYS_SUITE="FV3_GFS_v15p2"'
  'FCST_LEN_HRS="48"'
  'LBC_SPEC_INTVL_HRS="6"'
  'DATE_FIRST_CYCL="20190615"'
  'DATE_LAST_CYCL="20190615"'
  'CYCL_HRS=( "00" )'
  'EXTRN_MDL_NAME_ICS="FV3GFS"'
  'EXTRN_MDL_NAME_LBCS="FV3GFS"'
  'USE_USER_STAGED_EXTRN_FILES="TRUE"'
  'FV3GFS_FILE_FMT_ICS="grib2"'
  'FV3GFS_FILE_FMT_LBCS="grib2"'
  'WTIME_RUN_FCST="01:00:00"'


Processing experiment "RUN_ENVIR="community"" ...

ERROR:
  From script:  "run_experiments.sh"
  Full path to script:  "/lustre/f2/dev/esrl/Dom.Heinzeller/ufs-srweather-app/ufs-srweather-app/regional_workflow/tests/run_experiments.sh"
The experiment/workflow configuration file (baseline_config_fp) for the
specified baseline (baseline_name) does not exist:
  baseline_name = "RUN_ENVIR="community""
  baseline_config_fp = "/lustre/f2/dev/esrl/Dom.Heinzeller/ufs-srweather-app/ufs-srweather-app/regional_workflow/tests/baseline_configs/config.RUN_ENVIR="community".sh"
Please correct and rerun.
Exiting with nonzero status.
```

**Comment by climbfuji** [comment]
Everyone, I am not sure if I should be running the GST end to end tests as part of this PR. Can someone else do that, please, once this PR and the app PR got merged (before the code freeze tomorrow)?

File: ush/config_defaults.sh [hunk 6]
```diff
@@ -61,6 +61,16 @@ RUN_ENVIR="nco"
 # machine-dependent value.  This is not used if SCHED is not set to 
 # "slurm".
 #
+# CLUSTERS_DEFAULT:
+# If using the slurm job scheduler (i.e. if SCHED is set to "slurm"), 
+# the default clusters to which to submit workflow tasks.  If a task 
+# does not have a specific variable that specifies the partition to which 
+# it will be submitted (e.g. CLUSTERS_HPSS, CLUSTERS; see below),
```
Comment (gsketefian) [inline ush/config_defaults.sh]: Append "_FCST" to "CLUSTERS", I think.

File: ush/config.gaea.sh (hunk 5 above)
Comment (gsketefian) [reply ush/config.gaea.sh]: If interested in retrying, here's how to run the WE2E tests:

1) In `regional_workflow/tests directory`, create a text file that contains the names of the WE2E tests you want to run, one per line.  Let's call the file `my_expts.txt`.  In this case, there's only one experiment to run, so this file would contain just one line that says:
`GST_release_public_v1`
2) Run the `run_experiments.sh` script as follows:
`run_experiments.sh expts_file=my_expts.txt machine=gaea account="esrl_bmcs"`
The basic configuration file for the GST test is `config.GST_release_public_v1.sh` in the `baseline_configs` subdirectory.  `run_experiments.sh` will take that basic configuration file and add more settings to it to obtain a final experiment configuration file called `config.sh` file (a copy of which is saved in the experiment directory).  It then runs the `generate_FV3LAM_wflow.sh` script to generate an experiment directory and, by default, places a cron job in your cron table to launch the rocoto workflow every 2 minutes.  By default, the experiment directory will be at `${HOMErrfs}/expt_dirs/GST_release_public_v1`, where `${HOMErrfs}` is the directory one above your clone of `ufs-srweather-app`.

In `run_experiments.sh`, you're going to encounter error messages that say certain variables are not set for Gaea.  Those will need to be set first before the script can finish successfully.

Comment (climbfuji) [reply ush/config.gaea.sh]: Thanks, that worked. Last thing I will change before we merge this PR is to install rocoto in a shared location (thus far using Chris Harrop's test version) and create a module for it to use.

Comment (gsketefian) [reply ush/config.gaea.sh]: Great!  If you made changes to run_experiments.sh, feel free to incorporate them into this PR.

Comment (climbfuji) [reply ush/config.gaea.sh]: Ok, done. I made the changes wrt rocoto as well, and the regional_workflow python environment is now in a central location. This PR and the associated PR in the srw app are good to go.

**Review by gsketefian — APPROVED** [review]
Looks good!

**Review by JeffBeck-NOAA — APPROVED** [review]
Tested successfully on Hera.

**Pull request merged**